index.js

```diff
@@ -1,77 +1,184 @@
+const Xache = require('xache')
+const safetyCatch = require('safety-catch')
 
-var isLinux = os.platform() === 'linux' // native recursive watching not supported here
-var watchDirectory = isLinux ? watchFallback : watchRecursive
+// Native recursive watching is not supported on Linux based platforms
+const isLinux = process.platform === 'linux' || process.platform === 'android'
+const watchDirectory = isLinux ? watchFallback : watchRecursive
 
-module.exports = watch
+module.exports = class RecursiveWatch {
+  constructor (filename, onchange) {
+    this.filename = filename
+    this.onchange = onchange
 
-function watch (name, onchange) {
-  var clear = null
-  var stopped = false
+    this.unwatch = null
 
-  fs.lstat(name, function (_, st) {
-    if (!st || stopped) {
-      stopped = true
-      return
+    this.opened = false
+    this.closed = false
+
+    this._opening = this._ready()
+    this._closing = null
+
+    this._opening.catch(safetyCatch)
+  }
+
+  ready () {
+    return this._opening
+  }
+
+  async _ready () {
+    const st = await fs.promises.lstat(this.filename)
+
+    if (this.closed) return
+
+    const watch = st.isDirectory() ? watchDirectory : watchFile
+    this.unwatch = watch(this.filename, this.onchange)
+
+    this.opened = true
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    if (!this.opened) await this._opening.catch(safetyCatch)
+
+    if (this.unwatch) {
+      await this.unwatch()
+      this.unwatch = null
     }
-    clear = st.isDirectory() ? watchDirectory(name, onchange) : watchFile(name, onchange)
-  })
+  }
+}
+
+class Watch {
+  constructor (filename, opts, onchange) {
+    this.filename = filename
+    this.opts = opts || {}
+    this.onchange = onchange
+
+    this.opened = true
+    this.closed = false
 
-  return function () {
-    if (stopped) return
-    stopped = true
-    if (clear) {
-      clear()
+    this._closing = null
+
+    this.watcher = fs.watch(this.filename, this.opts, this.onchange)
+    this.watcherClosed = false
+
+    this.watcher.on('error', noop)
+    this.watcher.on('close', () => {
+      this.watcherClosed = true
+    })
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    let onclose = null
+
+    if (!this.watcherClosed) {
+      onclose = new Promise(resolve => this.watcher.on('close', resolve))
     }
+
+    this.watcher.close()
+    await onclose
+
+    this.watcher = null
   }
 }
 
 function watchFile (filename, onchange) {
-  var prev = null
-  var prevTime = 0
+  let prev = null
+  let prevTime = 0
+  let actives = 0
+
+  let closed = false
+  let close = null
+  const closing = new Promise(resolve => { close = resolve })
```

Comment:
It needs to be created at the beginning, it was part of a quick fix for the cleanup leaks

Those properties are not exposed so they don't necessarily need to match the same name convention and logic as ReadyResource, I agree it would look way better in a class so we can actually follow that